File: sandbox-runner/patterns/resumable-activity/python/activities.py
from temporalio import activity
from temporalio.exceptions import ApplicationError

from shared import TransferInput

# Valid accounts for the demo — any other account number fails.
VALID_ACCOUNTS = {"account-123", "account-456", "account-789"}


@activity.defn
async def execute_transfer(transfer: TransferInput) -> None:
    info = activity.info()
    activity.logger.info(
        f"Executing transfer: {transfer.amount} from {transfer.from_account} "
        f"to {transfer.to_account} (attempt {info.attempt})"
    )

    if transfer.to_account not in VALID_ACCOUNTS:
        raise ApplicationError(
            f"Account {transfer.to_account} not found",
            type="AccountNotFoundError",
            non_retryable=True,
        )

    activity.logger.info(
        f"Transfer of {transfer.amount} to {transfer.to_account} completed"
    )

File: sandbox-runner/patterns/resumable-activity/typescript/activities.ts
import { activityInfo, ApplicationFailure } from "@temporalio/activity";

import type { TransferInput } from "./shared";

// Valid accounts for the demo — any other account number fails.
const VALID_ACCOUNTS = new Set(["account-123", "account-456", "account-789"]);

export async function executeTransfer(transfer: TransferInput): Promise<void> {
  const { attempt } = activityInfo();
  console.log(
    `Executing transfer: ${transfer.amount} from ${transfer.fromAccount} ` +
      `to ${transfer.toAccount} (attempt ${attempt})`,
  );

  if (!VALID_ACCOUNTS.has(transfer.toAccount)) {
    throw ApplicationFailure.nonRetryable(
      `Account ${transfer.toAccount} not found`,
      "AccountNotFoundError",
    );
  }

  console.log(
    `Transfer of ${transfer.amount} to ${transfer.toAccount} completed`,
  );
}
